7080 multi select on aspects on mobile

tclaus opened the pull request:
Fixes Issue #7080 and updating Pr #7602 from sebastienadam .

Multiple aspects in mobile views. 

Moved Aspects in mobile view in one line with share and upload photos button, as in desktop view. 
This is what aspects selector look on mobile now: 
<img width="727" alt="Bildschirmfoto 2021-03-06 um 20 27 39" src="https://user-images.githubusercontent.com/501326/110232003-3120a680-7f1b-11eb-9f41-a1d3d02b87fa.png">


**Notes on failing Styling Tests**: The one styling was about renaming a -style issue- to change _ to - and a follow-up not to use IDs at general, so this styling issue was there forever. 

**Notes on Tests**: There are timeouts, but the cases run green. No Idea what to do about this.

sebastienadam committed:
#7080 using aspect selecting button form html into mobile version

sebastienadam committed:
#7080 Adaptation of the JS for the behavior of the aspect selection button.

sebastienadam committed:
#7080: SCSS adaptation for displaying the selection of the aspect selection button

tclaus committed:
Finishing Touches

diaspora-code-review commented on app/assets/stylesheets/mobile/mobile.scss:
```diff
@@ -658,39 +658,60 @@ form#update_profile_form {
 }
 
-#publisher_mobile {
+#publisher-mobile {
```
Avoid using id selectors

tclaus replied:
Old existing Selector - changed '_ 'to '-'

Flaburgan replied:
Yes you can ignore this one.

tclaus commented:
@Flaburgan Whats about the time-out tests? Can these be triggered again? Is it something what has its cause in the code?

Flaburgan commented:
> @Flaburgan Whats about the time-out tests? Can these be triggered again? Is it something what has its cause in the code?

I was checking that yesterday, when we were using Travis I was able to re-run builds, but looks like I can't do that with Github Actions, I don't know if that's a permission problem or a feature not available, can you tell @SuperTux88 ?

SuperTux88 commented:
I can re-trigger it, but if all 4 cucumber tests failed with the same test, it's probably a real error that doesn't go away with re-running it, so there is no point of triggering the tests again with the current state (and the timeout problem with jasmine can just be ignored for this run, it will re-run again on next commit anyway).

tclaus commented:
> , it will re-run again on next commit anyway).

OK, I will check this again. Thanks for the analysis.

tclaus commented:
All tests runs well locally - so if any of these still fail, I request some help.

diaspora-code-review commented on features/step_definitions/web_steps.rb:
```diff
@@ -31,7 +31,7 @@ def with_scope(locator)
 When /^(?:|I )press "([^"]*)"(?: within "([^"]*)")?$/ do |button, selector|
   with_scope(selector) do
-    click_button(button)
+    find_button(button).trigger('click')
```
Style/StringLiterals: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.

diaspora-code-review commented on features/support/env.rb:
```diff
@@ -23,8 +23,13 @@
 
+options = {
+  window_size: [1920, 1080],
+  timeout: 30
```
Layout/HashAlignment: Align the keys and values of a hash literal if they span more than one line.

diaspora-code-review commented on features/support/env.rb:
```diff
@@ -23,6 +23,11 @@
 
 Capybara.server = :webrick
 
+options = {
```
Lint/UselessAssignment: Useless assignment to variable - `options`.

Flaburgan commented on features/mobile/posts_from_main_page.feature:
```diff
@@ -21,8 +21,10 @@ Feature: posting from the mobile main page
-    And I select "Unicorns" from "aspect_ids_"
-    And I press "Share"
+    And I press the aspect dropdown
+    And I toggle the aspect "Unicorns"
```
Could you please add a test where you post to two different aspects and check that the post is visible in these two and only these two?

tclaus replied:
Oui.

tclaus replied:
@Flaburgan Can it be that there is no test for setting more than one aspect at same time? Even in Desktop-Mode?

Flaburgan replied:
They are: https://github.com/diaspora/diaspora/blob/develop/features/desktop/posts_from_main_page.feature#L58

tclaus replied:
Help requested - cucumber tests still won't work. I don't get test to work - the Aspects button will (in test) not expand, or aspects are not 'visible' However - on real browser everything seems to work.

Flaburgan reviewed (changes requested):
Great job, it is working well and the design is nice. Only a comment about the test coverage and then we should be good to go.

jhass commented on features/step_definitions/aspects_steps.rb:
```diff
@@ -112,6 +116,12 @@ def aspect_dropdown_visible?
+When /^I press the share button$/ do
+  # There were issues 'clicking' the share button on mobile
+  click_share_button
+  sleep 1 # Wait a little bit - especially on mobile views shaing dont works every time
```
Ideally there's some UI change we can observe and wait for, something like `expect(page).to have_selector("#dropdown.visible")`, did you have any luck with something like this? :)

jhass commented on app/views/aspects/_aspect_dropdown.mobile.haml:
```diff
@@ -0,0 +1,36 @@
+  if current_user.getting_started?
+    dropdown_css[:title] = popover_with_close_html("2. #{t('shared.public_explain.control_your_audience')}")
+    dropdown_css["data-content"] = t("shared.public_explain.visibility_dropdown")
+  end
```
I went through the sign up flow on mobile and even reaching the publisher page already had `getting_started` be `false` on my new user. Manually setting it back to `true` showed for this to have no effect. Let's drop this.

tclaus commented:
@Flaburgan @jhass  I added cucumber tests. 
Note: I went into serous problems creating and running these tests because I had to learn to run `rake assets:precompile ` whenever I change templates before running cucumber tests! Cucumber then runs on something 'old'.

tclaus commented:
Ping @jhass / @Flaburgan 
I added tests and fixed code.  Can you approve it now?
(The whining pronto/scss is a "Avoid using id selectors" which is from the legacy code and built more deep into architecture to change to in passing by)

tclaus commented:
Issues are solved, tests added, everything is green. Is there anything more to do for an approval? 
@Flaburgan , @jhass

SuperTux88 reviewed (changes requested):
There are some mostly minor things I found, but I think it should be quick to fix them. Can you please also squash some of the commits together (style fixes, cleanup and reverts should just be squashed with the commit where it was initially changed, so we only end up with commits with actual changes and meaningful commit messages).

Sidenote: this doesn't fix #7333 yet, and that is fine and can be done in a separate PR. But I think it should be fixed in the same release as this one gets merged, otherwise it will be even more confusing for users as it already is.

tclaus commented on app/assets/stylesheets/mobile/mobile.scss:
```diff
@@ -658,40 +658,62 @@ form#update_profile_form {
   .submit_block { margin-bottom: 20px; }
 }
 
-select#user_language, select#user_color_theme, #user_auto_follow_back_aspect_id, #aspect_ids_ {
-  padding: 3px;
-}
-
```
I have added this again, except the aspect_ids_. A bad way to define styles for programmatic built Ids.. I found the f.select .. part which builds this Ids

SuperTux88 replied:
Not really the scope of this PR, but maybe it can be replaced by just a class then instead, which then doesn't need to be programmatically built and can be easier found? Or just select all `select` elements on the settings page (no idea why it's only these specific elements)? :thinking:

tclaus replied:
I changed it to class settings. But I do not "see" any change when a select element has a padding.

SuperTux88 replied:
With the `padding: 3px`:
![image](https://user-images.githubusercontent.com/458548/188008868-03c3ffd6-54c0-4c7d-9363-e52936031fe4.png)
Without the `padding: 3px`:
![image](https://user-images.githubusercontent.com/458548/188008625-996d165c-b521-458d-8fcf-ac02d1b666f4.png)

tclaus replied:
Ah, now I see it. Thanks.

tclaus commented on app/assets/stylesheets/mobile/mobile.scss:
```diff
@@ -658,40 +658,62 @@ form#update_profile_form {
+@mixin aspect-dropdown-link() {
+  $link-text-color: #333;
+
+  a {
+    cursor: pointer;
+    padding-left: 10px;
+
+    .text {
+      color: $link-text-color;
+      font-size: 128%;
+    }
+
+    &:hover {
+      background: $background-grey;
+    }
+  }
+}
```
This is always inside a mixin. I don't change it (for know)

SuperTux88 replied:
Yes, it's a mixin, but it's almost the same mixin defined twice, and both time it's only used once. Defining it once and using it twice would be better (and the one little difference can then just be added as an override where it needs to be different).

tclaus replied:
@SuperTux88 I moved this into a nice mixin.

tclaus committed:
Using mixins for aspects css style

SuperTux88 commented on app/assets/stylesheets/mobile/settings.scss:
```diff
@@ -57,3 +57,9 @@
   }
 }
+
+.user-language,
+.user-color-theme,
+.aspect-id {
```
With "replaced by just a class" I meant using one class for all of them together, so you don't need a separate class for each element. Like for example `.settings-dropdown` or something like that. Or as said, it probably also works if you just select all the `select` elements on the settings page, so no extra class is needed at all? That way it would also automatically work if for some reason we would add another dropdown which should also be styled that way on mobile.

```suggestion
.settings-container select {
```

(or better move it with only `select` inside the `.settings-container` block above :) )

tclaus replied:
I now have set the property to all selects inside the settings-container

SuperTux88 reviewed (changes requested):
Looks good :+1: only one last little thing.

SuperTux88 commented on app/views/users/_edit.haml:
```diff
@@ -71,7 +71,7 @@
         = form_for "user", url: edit_user_path, html: {method: :put} do |f|
           .form-inline.clearfix
-            = f.select :color_theme, available_color_themes, {}, class: "form-control form-group"
+            = f.select :color_theme, available_color_themes, {}, class: "form-control form-group color-theme"
```
There is still a leftover class here.

tclaus replied:
Removed, thanks for being such an eagle-eye.

SuperTux88 reviewed (approved):
Looks good now :cookie:
